src/phalanx/models/secrets.py

```diff
@@ -0,0 +1,237 @@
+"""Pydantic models for Phalanx application secrets."""
+
+from __future__ import annotations
+
+from enum import Enum
+from typing import Any, Self
+
+from pydantic import BaseModel, Extra, Field, SecretStr, validator
+
+__all__ = [
+    "RequiredSecret",
+    "RequiredSecretConfig",
+    "ResolvedSecret",
+    "Secret",
+    "SecretConfig",
+    "SecretCopyRules",
+    "SecretGenerateRules",
+    "SecretGenerateType",
+]
+
+
+class SecretCopyRules(BaseModel):
+    """Rules for copying a secret value from another secret."""
+
+    application: str
+    """Application from which the secret should be copied."""
+
+    key: str
+    """Secret key from which the secret should be copied."""
+
+    condition: str | None = Field(
+        None,
+        description=(
+            "Helm chart value that, if set, indicates the secret should be"
+            " copied"
+        ),
+        alias="if",
+    )
+
+    class Config:
+        allow_population_by_field_name = True
+        extra = Extra.forbid
+
+
+class SecretGenerateType(Enum):
+    """Type of secret for generated secrets."""
+
+    password = "password"
+    gafaelfawr_token = "gafaelfawr-token"
+    fernet_key = "fernet-key"
+    rsa_private_key = "rsa-private-key"
+    bcrypt_password_hash = "bcrypt-password-hash"
+    mtime = "mtime"
+
+
+class SecretGenerateRules(BaseModel):
+    """Rules for generating a secret value."""
+
+    type: SecretGenerateType
+    """Type of secret."""
+
+    source: str | None = None
+    """Key of secret on which this secret is based.
+
+    This may only be set by secrets of type ``bcrypt-password-hash`` or
+    ``mtime``.
+    """
+
+    condition: str | None = Field(
+        None,
+        description=(
+            "Helm chart value that, if set, indicates the secret should be"
+            " generated"
+        ),
+        alias="if",
+    )
+
+    class Config:
+        allow_population_by_field_name = True
+        extra = Extra.forbid
+
+    @validator("source")
+    def _validate_source(
+        cls, v: str | None, values: dict[str, Any]
+    ) -> str | None:
+        secret_type = values["type"]
+        want_value = secret_type in (
+            SecretGenerateType.bcrypt_password_hash,
+            SecretGenerateType.mtime,
+        )
+        if v is None and want_value:
+            msg = f"source not set for secret of type {secret_type}"
+            raise ValueError(msg)
+        if v is not None and not want_value:
+            msg = f"source not allowed for secret of type {secret_type}"
+            raise ValueError(msg)
+        return v
+
+
+class RequiredSecretConfig(BaseModel):
```

Comment:
I'll say it took my brain a minute to grok the hierarchy of models, but I don't think  have any recommendations. Somehow making `SecretConfig` a subclass of `RequiredSecretConfig` and then making `RequiredSecret` and subclass of `SecretConfig` threw me 😂

Reply:
Yeah, this was bothering me too. I just reworked the models to hopefully clear this up and to not resolve conditions in more than one place, and I think the result is a bit cleaner.